libp2p/protocols/kademlia/lookupstate.nim

```diff
@@ -0,0 +1,120 @@
+import ./consts
+import ./protobuf
+import ./xordistance
+import ./keys
+import ../../[peerid, peerinfo]
+import algorithm
+import chronicles
+
+type
+  LookupNode* = object
+    peerId: PeerId
+    distance: XorDistance
+    queried: bool # have we already queried this node?
+    pending: bool # is there an active request rn?
+    failed: bool # did the query timeout or error?
+
+  LookupState* = object
+    targetId: Key
+    shortlist: seq[LookupNode] # current known closest node
+    activeQueries*: int # how many queries in flight
+    alpha: int # parallelism level
+    repliCount: int ## aka `k` in the spec: number of closest nodes to find
+    done*: bool # has lookup converged
+
+proc alreadyInShortlist(state: LookupState, peer: Peer): bool =
+  for p in state.shortlist:
+    if p.peerId.getBytes() == peer.id:
+      return true
+  return false
```

Comment:
We could just return a `filterIt(...).len == 0`, no?

Reply:
```nim
proc alreadyInShortlist(state: LookupState, peer: Peer): bool =
  return state.shortlist.anyIt(it.peerId.getBytes() == peer.id)
```
Will use this.